_posts/2019-05-20-server-rendering-responsively.md

```diff
@@ -0,0 +1,201 @@
+toolbox. Here's a component that would render a `div` (`Box`) with a width of 50% at our mobile breakpoint, 75% at
+our tablet breakpoint, and 100% for anything larger:
+
+```xml
+<Box width={["50%", "75%", "100%"]}>
+  ...
+</Box>
+```
+
+Here's a real example
+[from our codebase](https://github.com/artsy/reaction/blob/f3dabb884d616c5c42bbd47b1432dc2a9456b8ca/src/Apps/Search/Components/SearchResultsSkeleton/Header.tsx#L6).
+
+```xml
+<Box height={100} mb={30} mt={120} pr={[0, 20]} pl={[0, 20]}>
+  <Box pl={[20, 0]}>
+    <Box mb={40} background={color("black10")} width={260} height={20} />
+    ...
+  </Box>
+</Box>
+```
+
+There are a lot of abbreviations in there - `mb` sets the bottom margin, `mt` sets the top margin, `pl` sets the
+left padding, and `pr` sets the right padding. And while only a few of the properties in this example _are_
+specifying an array of sizes to be used at different breakpoints, all of those properties _can_ take an array for
+different breakpoints. As developers, we love this experience. We can make subtle differences to components across
+breakpoints with very little code and effort.
+
+We use `styled-system` extensively within [our design system](https://palette.artsy.net/). You can
+[poke around our source](https://github.com/artsy/reaction/blob/f3dabb884d616c5c42bbd47b1432dc2a9456b8ca/src/Apps/Search/Components/SearchResultsSkeleton/Header.tsx#L6)
+to see how much we've embraced `styled-system`'s responsive styles.
+
+There's one type of challenge with building a responsive app that `styled-system` can't solve: when we need to emit
+different layouts across different breakpoints. In this case, we need something that can render very different
+component sub-trees. We couldn't find an approach that satisfied our needs, so we wrote our own.
+
+## Tool 2: [`@artsy/react-responsive-media`](https://github.com/artsy/react-responsive-media)
+
+First off, an announcement: we've just released `react-responsive-media` version 1.0! (TODO: release v1.0 😬 or
+change this announcement)
+
+`react-responsive-media` allows you to define a set of breakpoint widths, then declaratively render component
+sub-trees when those breakpoints are met. It looks something like this:
+
+```xml
+<>
+  <Media at="xs">
+    <MobileLayout />
+  </Media>
+  <Media greaterThan="xs">
+    <NonMobileLayout />
+  </Media>
+</>
+```
+
+In this example, we're emitting the `MobileLayout` component for devices at or below our `xs` breakpoint, and the
+`NonMobileLayout` for devices greater than our `xs` breakpoint. You can imagine that the `MobileLayout` and
+`NonMobileLayout` components contain complicated sub-trees, with more significant differences than `styled-system`
+could handle.
+
+### How it works
+
+The first important thing to note is that when server-rendering with `react-responsive-media`, **all** breakpoints
+get rendered by the server.
```

Comment:
> …and `Media` components are wrapped by plain CSS that will only show them on the client if the breakpoint matches the user’s current browser size. This means that the client can accurately start rendering the HTML/CSS _while_ it receives it, which is long before the React application has booted and thus leads to improved visual end-user performance.

Reply:
(Suggested in https://github.com/artsy/artsy.github.io/pull/566/files#r286231699)